### PR TITLE
Utf8 Component

### PR DESCRIPTION
**Pull request by hhamon**

### CODE REVIEW

File: src/Symfony/Component/Utf8/Exception/InvalidArgumentException.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Utf8\Exception;
+
+class InvalidArgumentException extends \InvalidArgumentException
```
Comment (fabpot) [inline src/Symfony/Component/Utf8/Exception/InvalidArgumentException.php]: You should create an ExceptionInterface that acts as a marker for all exceptions coming from this component as done by other components.

File: src/Symfony/Component/Utf8/GenericStringInterface.php [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    public function startsWith(string $prefix, int $offset = 0): bool;
+
+    /**
+     * Returns whether or not the string ends with the given suffix.
+     *
+     * @param string $suffix The suffix to test
+     *
+     * @return bool
+     */
+    public function endsWith(string $suffix): bool;
```
Comment (fabpot) [inline src/Symfony/Component/Utf8/GenericStringInterface.php]: Is requiring PHP7 done on purpose?

File: src/Symfony/Component/Utf8/String.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
+String Class APIs Comparison
```
Comment (fabpot) [inline src/Symfony/Component/Utf8/String.md]: Should probably be removed

File: src/Symfony/Component/Utf8/Exception/InvalidArgumentException.php (hunk 1 above)
Comment (hhamon) [reply src/Symfony/Component/Utf8/Exception/InvalidArgumentException.php]: Fixed.

File: src/Symfony/Component/Utf8/String.md (hunk 3 above)
Comment (hhamon) [reply src/Symfony/Component/Utf8/String.md]: Done.

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        if ($maxChunkLength < 1) {
+            throw new InvalidArgumentException('The maximum length of each segment must be greater than zero.');
+        }
+
+        if ($maxChunkLength > 65535) {
+            $maxChunkLength = 65535;
+        }
+
+        if ('' === $this->string) {
+            yield clone $this;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: just "return"? str_split return `array('')` with empty strings, but I don't get why this is useull at all.
WDYT? (same in the other "getIterator")

File: src/Symfony/Component/Utf8/GenericStringInterface.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    public static function compareNaturalIgnoreCase(string $a, string $b): int;
+
+    /**
+     * @return static
+     */
+    public function reverse();
+
+    public function length(): int;
+
+    public function width(): int;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/GenericStringInterface.php]: we'll need an arg here, to allow for non-countoing for CLI decorations - could even be the default to me

File: src/Symfony/Component/Utf8/GenericStringInterface.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    /**
+     * @return static
+     */
+    public function toLowerCase();
+
+    /**
+     * @return static
+     */
+    public function toUpperCase();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/GenericStringInterface.php:84]: let's add another one: toFoldedCase, could be useful to me

File: src/Symfony/Component/Utf8/GenericStringInterface.php [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+    public function append(string $suffix);
+
+    public function prepend(string $prefix);
+
+    public function replace($from, $to, int $offset = 0, int $length = null, int &$count = null);
+
+    public function replaceIgnoreCase($from, $to, int $offset = 0, int $length = null, int &$count = null);
+
+    public function toUpperCaseFirst($allWords = false);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/GenericStringInterface.php]: should be moved near to*Case

File: src/Symfony/Component/Utf8/Bytes.php [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+
+        return $result;
+    }
+
+    public function width(): int
+    {
+        return strlen($this->string);
+    }
+
+    public function replace($from, $to, int $offset = 0, int $length = null, int &$count = null)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: hum, not sure about the meaning of $offset / $length anymore :)

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+{
+    private $string = '';
+
+    public static function fromString(string $string): self
+    {
+        if (!preg_match('//u', $string)) {
+            throw new InvalidArgumentException('Given string is not a valid UTF-8 encoded string.');
+        }
+
+        $codePoints = new self($string);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: there no constructor, thus no need to pass any arg

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trim(string $charsList = null): self
+    {
+        $charsList = $charsList ? preg_quote($charsList) : '[:space:]';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: missing utf8 validity check for charsList

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trimLeft(string $charsList = null): self
+    {
+        $charsList = $charsList ? preg_quote($charsList) : '[:space:]';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: utf8 check

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trimRight(string $charsList = null): self
+    {
+        $charsList = $charsList ? preg_quote($charsList) : '[:space:]';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: charsList

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+        if ($maxChunkLength < 1) {
+            throw new InvalidArgumentException('The maximum length of each segment must be greater than zero.');
+        }
+
+        if ('' === $this->string) {
+            return;
+        }
+
+        if ($maxChunkLength > 65535) {
+            $maxChunkLength = 65535;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: throw instead?

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+        }
+
+        foreach (preg_split('/(.{'.$maxChunkLength.'})/u', $this->string, -1, PREG_SPLIT_DELIM_CAPTURE | PREG_SPLIT_NO_EMPTY) as $key => $char) {
+            $clone = clone $this;
+            $clone->string = $char;
+            yield $key => $clone;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: no need for $key

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 15]
```diff
@@ -0,0 +1,390 @@
...
+        }
+    }
+
+    public function append(string $suffix): self
+    {
+        $result = clone $this;
+        $result->string.= self::fromString($suffix)->string;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: missing space before .=

Comment (nicolas-grekas) [reply src/Symfony/Component/Utf8/CodePoints.php]: let's to the check inline, no need to call the static constructor

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 16]
```diff
@@ -0,0 +1,390 @@
...
+
+        return $result;
+    }
+
+    public function prepend(string $prefix): self
+    {
+        $result = self::fromString($prefix);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: utf8 check inline

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 17]
```diff
@@ -0,0 +1,390 @@
...
+        $result->string.= $this->string;
+
+        return $result;
+    }
+
+    public function toBytes(): Bytes
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: add this (and the two others) to the interface? toCodePoints would just return $this here

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 18]
```diff
@@ -0,0 +1,390 @@
...
+        return Graphemes::fromString($this->string);
+    }
+
+    public function indexOf(string $needle, int $offset = 0)
+    {
+        if ('' === $needle) {
+            return null;
+        }
+
+        if (PHP_VERSION_ID < 70100 && $offset < 0) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: the look strange, the behavior should no depend on the phpversion

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 19]
```diff
@@ -0,0 +1,390 @@
...
+        return false === $result ? null : $result;
+    }
+
+    public function indexOfIgnoreCase(string $needle, int $offset = 0)
+    {
+        if ('' === $needle) {
+            return null;
+        }
+
+        if (PHP_VERSION_ID < 70100 && $offset < 0) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: php version behavior dependency?

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 20]
```diff
@@ -0,0 +1,390 @@
...
+            return null;
+        }
+
+        $result = mb_stripos($this->string, $needle, $offset);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: no "> 0" check?

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 21]
```diff
@@ -0,0 +1,390 @@
...
+
+    public function substringOfIgnoreCase(string $needle, int $offset = 0)
+    {
+        if ('' === $needle) {
+            return null;
+        }
+
+        $haystack = $offset > 0 ? mb_substr($this->string, $offset, null, 'UTF-8') : $this->string;
+
+        $result = mb_stristr($haystack, $needle, null, 'UTF-8');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: 3d arg should be "false"

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+
+    public function substringOf(string $needle, int $offset = 0)
+    {
+        if ('' === $needle) {
+            return null;
+        }
+
+        $haystack = $offset > 0 ? mb_substr($this->string, $offset, null, 'UTF-8') : $this->string;
+
+        $result = mb_strstr($haystack, $needle, null, 'UTF-8');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: 3d arg should be "false"

Comment (nicolas-grekas) [reply src/Symfony/Component/Utf8/CodePoints.php]: should be $beforeNeedle, new arg on the mthod in fact, isn't it? (same for methods below)

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 23]
```diff
@@ -0,0 +1,390 @@
...
+    }
+
+    public static function compare(string $a, string $b): int
+    {
+        return strcmp($a, $b);
+    }
+
+    public static function compareIgnoreCase(string $a, string $b): int
+    {
+        return strcasecmp($a, $b);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: bad implem for utf8, should use mbstring at some point (same below)

File: src/Symfony/Component/Utf8/CodePoints.php [hunk 24]
```diff
@@ -0,0 +1,390 @@
...
+    {
+        $result = new self();
+        $result->string = implode('', array_reverse(iterator_to_array($this)));
+
+        return $result;
+    }
+
+    public function width(): int
+    {
+        return mb_strwidth($this->string, 'UTF-8');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/CodePoints.php]: better take the implem of patchwork here

File: src/Symfony/Component/Utf8/Resources/data/caseFolding_full.ser [hunk 25]
```diff
@@ -0,0 +1 @@
...5:"ἂι";i:20;s:5:"ἃι";i:21;s:5:"ἄι";i:22;s:5:"ἅι";i:23;s:5:"ἆι";i:24;s:5:"ἇι";i:25;s:5:"ἀι";i:26;s:5:"ἁι";i:27;s:5:"ἂι";i:28;s:5:"ἃι";i:29;s:5:"ἄι";i:30;s:5:"ἅι";i:31;s:5:"ἆι";i:32;s:5:"ἇι";i:33;s:5:"ἠι";i:34;s:5:"ἡι";i:35;s:5:"ἢι";i:36;s:5:"ἣι";i:37;s:5:"ἤι";i:38;s:5:"ἥι";i:39;s:5:"ἦι";i:40;s:5:"ἧι";i:41;s:5:"ἠι";i:42;s:5:"ἡι";i:43;s:5:"ἢι";i:44;s:5:"ἣι";i:45;s:5:"ἤι";i:46;s:5:"ἥι";i:47;s:5:"ἦι";i:48;s:5:"ἧι";i:49;s:5:"ὠι";i:50;s:5:"ὡι";i:51;s:5:"ὢι";i:52;s:5:"ὣι";i:53;s:5:"ὤι";i:54;s:5:"ὥι";i:55;s:5:"ὦι";i:56;s:5:"ὧι";i:57;s:5:"ὠι";i:58;s:5:"ὡι";i:59;s:5:"ὢι";i:60;s:5:"ὣι";i:61;s:5:"ὤι";i:62;s:5:"ὥι";i:63;s:5:"ὦι";i:64;s:5:"ὧι";i:65;s:5:"ὰι";i:66;s:4:"αι";i:67;s:4:"άι";i:68;s:4:"ᾶ";i:69;s:6:"ᾶι";i:70;s:4:"αι";i:71;s:5:"ὴι";i:72;s:4:"ηι";i:73;s:4:"ήι";i:74;s:4:"ῆ";i:75;s:6:"ῆι";i:76;s:4:"ηι";i:77;s:6:"ῒ";i:78;s:6:"ΐ";i:79;s:4:"ῖ";i:80;s:6:"ῗ";i:81;s:6:"ῢ";i:82;s:6:"ΰ";i:83;s:4:"ῤ";i:84;s:4:"ῦ";i:85;s:6:"ῧ";i:86;s:5:"ὼι";i:87;s:4:"ωι";i:88;s:4:"ώι";i:89;s:4:"ῶ";i:90;s:6:"ῶι";i:91;s:4:"ωι";i:92;s:2:"ff";i:93;s:2:"fi";i:94;s:2:"fl";i:95;s:3:"ffi";i:96;s:3:"ffl";i:97;s:2:"st";i:98;s:2:"st";i:99;s:4:"մն";i:100;s:4:"մե";i:101;s:4:"մի";i:102;s:4:"վն";i:103;s:4:"մխ";}}
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Resources/data/caseFolding_full.ser]: should be turned to a plain php array

File: src/Symfony/Component/Utf8/Bytes.php [hunk 26]
```diff
@@ -0,0 +1,349 @@
...
+        $bytes->string = $string;
+
+        return $bytes;
+    }
+
+    public static function fromCharCode(int ...$codes): self
+    {
+        $string = '';
+        foreach ($codes as $code) {
+            $string.= chr($code);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: missing space

File: src/Symfony/Component/Utf8/Bytes.php [hunk 27]
```diff
@@ -0,0 +1,349 @@
...
+        $bytes = new self();
+        $bytes->string = $string;
+
+        return $bytes;
+    }
+
+    public static function fromCharCode(int ...$codes): self
+    {
+        $string = '';
+        foreach ($codes as $code) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: array_map('chr', ... ?

File: src/Symfony/Component/Utf8/Bytes.php [hunk 28]
```diff
@@ -0,0 +1,349 @@
...
+        return $chunks;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toLowerCase(): self
+    {
+        $result = clone $this;
+        $result->string = strtolower($this->string);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: we need to make this locale independent if possible (not I-to-turkish "i" suprize)

Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: we need to make this locale independent if possible (not I-to-turkish "i" surprise)

File: src/Symfony/Component/Utf8/Bytes.php [hunk 29]
```diff
@@ -0,0 +1,349 @@
...
+    {
+        if ($maxChunkLength < 1) {
+            throw new InvalidArgumentException('The maximum length of each segment must be greater than zero.');
+        }
+
+        if ('' === $this->string) {
+            return;
+        }
+
+        foreach (str_split($this->string, $maxChunkLength) as $key => $char) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: no need for $key

File: src/Symfony/Component/Utf8/Bytes.php [hunk 30]
```diff
@@ -0,0 +1,349 @@
...
+            $clone = clone $this;
+            $clone->string = $char;
+            yield $key => $clone;
+        }
+    }
+
+    public function append(string $suffix): self
+    {
+        $result = clone $this;
+        $result->string.= $suffix;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: missing space

File: src/Symfony/Component/Utf8/Bytes.php [hunk 31]
```diff
@@ -0,0 +1,349 @@
...
+        return Graphemes::fromString($this->string);
+    }
+
+    public function indexOf(string $needle, int $offset = 0)
+    {
+        if ('' === $needle) {
+            return null;
+        }
+
+        if (PHP_VERSION_ID < 70100 && $offset < 0) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: version check?

File: src/Symfony/Component/Utf8/Bytes.php [hunk 32]
```diff
@@ -0,0 +1,349 @@
...
+        return false === $result ? null : $result;
+    }
+
+    public function indexOfIgnoreCase(string $needle, int $offset = 0)
+    {
+        if ('' === $needle) {
+            return null;
+        }
+
+        if (PHP_VERSION_ID < 70100 && $offset < 0) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: version check?

File: src/Symfony/Component/Utf8/Bytes.php [hunk 33]
```diff
@@ -0,0 +1,349 @@
...
+    {
+        $result = new self();
+        $result->string = strrev($this->string);
+
+        return $result;
+    }
+
+    public function width(): int
+    {
+        return strlen($this->string);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Bytes.php]: should be ascii only version of patchwork's implem

File: src/Symfony/Component/Utf8/Graphemes.php [hunk 34]
```diff
@@ -0,0 +1,452 @@
...
+    {
+        $result = clone $this;
+        $result->string = grapheme_substr($this->string, $start, $length);
+
+        return $result;
+    }
+
+    public function trim(string $charsList = null): self
+    {
+        $charsList = $charsList ? preg_quote($charsList) : '[:space:]';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Graphemes.php]: utf8 check (same below)

File: src/Symfony/Component/Utf8/Graphemes.php [hunk 35]
```diff
@@ -0,0 +1,452 @@
...
+        if ($maxChunkLength < 1) {
+            throw new InvalidArgumentException('The maximum length of each segment must be greater than zero.');
+        }
+
+        if ('' === $this->string) {
+            return;
+        }
+
+        if ($maxChunkLength > 65535) {
+            $maxChunkLength = 65535;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Graphemes.php]: throw

File: src/Symfony/Component/Utf8/Graphemes.php [hunk 36]
```diff
@@ -0,0 +1,452 @@
...
+        }
+
+        $length = strlen($this->string);
+        $i = $key = 0;
+
+        while ($i < $length) {
+            $clone = clone $this;
+            $clone->string = grapheme_extract($this->string, $maxChunkLength, GRAPHEME_EXTR_COUNT, $i, $i);
+            yield $key++ => $clone;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Graphemes.php]: no need for $key

File: src/Symfony/Component/Utf8/Graphemes.php [hunk 37]
```diff
@@ -0,0 +1,452 @@
...
+        }
+    }
+
+    public function append(string $suffix): self
+    {
+        $result = clone $this;
+        $result->string.= self::fromString($suffix)->string;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Graphemes.php]: no static constructor (same below)

File: src/Symfony/Component/Utf8/Graphemes.php [hunk 38]
```diff
@@ -0,0 +1,452 @@
...
+        return CodePoints::fromString($this->string);
+    }
+
+    public function indexOf(string $needle, int $offset = 0)
+    {
+        if ('' === $needle) {
+            return null;
+        }
+
+        if (PHP_VERSION_ID < 70100 && $offset < 0) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Graphemes.php]: ver check? (same below)

File: src/Symfony/Component/Utf8/Bytes.php (hunk 28 above)
Comment (hhamon) [reply src/Symfony/Component/Utf8/Bytes.php]: What do you mean exactly?

File: src/Symfony/Component/Utf8/GenericStringInterface.php (hunk 5 above)
Comment (hhamon) [reply src/Symfony/Component/Utf8/GenericStringInterface.php]: like `$strict = true`?

File: src/Symfony/Component/Utf8/Graphemes.php [hunk 39]
```diff
@@ -0,0 +1,452 @@
...
+
+            if ($width < $count = iconv_strlen($string, 'UTF-8') + $wide + $count) {
+                $width = $count;
+            }
+        }
+
+        return $width;
+    }
+
+    public function replace(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Graphemes.php]: bad CS

File: src/Symfony/Component/Utf8/Graphemes.php [hunk 40]
```diff
@@ -0,0 +1,452 @@
...
+        $from,
+        $to,
+        int $offset = 0,
+        int $length = null,
+        int &$count = null
+    ) {
+        // TODO: Implement replace() method.
+    }
+
+    public function replaceIgnoreCase(
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Graphemes.php]: bad cs

File: src/Symfony/Component/Utf8/Graphemes.php [hunk 41]
```diff
@@ -0,0 +1,452 @@
...
+        $to,
+        int $offset = 0,
+        int $length = null,
+        int &$count = null
+    ) {
+        // TODO: Implement replaceIgnoreCase() method.
+    }
+
+    /**
+     * @return static
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Utf8/Graphemes.php]: inheritdoc everywhere?

**Comment by nicolas-grekas** [comment]
Continued in https://github.com/symfony/symfony/pull/33553

**Pull request closed without merging**